Firestore/core/src/firebase/firestore/immutable/llrb_node_iterator.h

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_LLRB_NODE_ITERATOR_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_IMMUTABLE_LLRB_NODE_ITERATOR_H_
+
+#include <iterator>
+#include <stack>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/immutable/llrb_node.h"
+#include "Firestore/core/src/firebase/firestore/util/comparison.h"
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace immutable {
+namespace impl {
+
+/**
+ * An iterator for traversing LlrbNodes.
+ *
+ * LlrbNode is an immutable tree, where insertions create new trees without
+ * invalidating any of the old instances. This means the tree cannot contain
+ * parent pointers and thus this iterator implementation must keep an explicit
+ * stack.
+ */
+template <typename N>
+class LlrbNodeIterator {
+ public:
+  using node_type = N;
+  using key_type = typename node_type::first_type;
+
+  using stack_type = std::stack<const node_type*>;
+
+  using iterator_category = std::forward_iterator_tag;
+  using value_type = typename node_type::value_type;
+
+  using pointer = typename node_type::value_type const*;
+  using reference = typename node_type::value_type const&;
+  using difference_type = std::ptrdiff_t;
+
+  /**
+   * Constructs an iterator starting at the first node in the iteration
+   * sequence of the tree represented by the given root node (i.e. it points at
+   * the left-most node).
+   */
+  static LlrbNodeIterator Begin(const node_type* root) {
+    stack_type stack;
+
+    const node_type* node = root;
+    while (!node->empty()) {
+      stack.push(node);
+      node = &node->left();
+    }
+
+    return LlrbNodeIterator{std::move(stack)};
+  }
+
+  /**
+   * Constructs an iterator pointing at the end of the iteration sequence of the
+   * tree pointed to by the given node (i.e. one past the right-most node)
+   */
+  static LlrbNodeIterator End() {
+    return LlrbNodeIterator{stack_type{}};
+  }
+
+  /**
+   * Returns true if this iterator points at the end of the iteration sequence.
+   */
+  bool end() const {
```

Comment:
How about `at_end`, `is_end` or similar? I find the name to be too similar to container's `end` member function, which does something different.